File: standalone-container/src/test/scala/com/yahoo/container/standalone/StandaloneContainer.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone

import com.yahoo.config.model.producer.AbstractConfigProducerRoot
import com.yahoo.config.model.test.MockRoot
import com.yahoo.container.Container
import com.yahoo.jdisc.test.TestDriver
import scala.xml.Node
import com.yahoo.vespa.model.VespaModel
import com.yahoo.io.IOUtils
import java.nio.file.{Files, Path}
import com.yahoo.vespa.model.container.xml.ContainerModelBuilder.Networking

/**
 * Creates a local application from vespa-services fragments.
 *
 * @author  tonytv
 */
object StandaloneContainer {
  def firstContainerId(root: AbstractConfigProducerRoot): String = {
    root.getConfigProducer("container").get().getConfigId
  }

  def withStandaloneContainer[T](containerNode: Node) {
    withTempDirectory { applicationDirectory =>
      System.setProperty(StandaloneContainerApplication.applicationLocationInstallVariable, applicationDirectory.toString)
      createServicesXml(applicationDirectory, containerNode)

      val driver = TestDriver.newInjectedApplicationInstance(classOf[StandaloneContainerApplication])
      driver.close()
      Container.resetInstance()
    }
  }

  def withContainerModel[T](containerNode: Node)(f: VespaModel => T) {
    withTempDirectory { applicationPath =>
      createServicesXml(applicationPath, containerNode)

      val distributedFiles = new LocalFileDb(applicationPath)
      val (root, container) = StandaloneContainerApplication.createContainerModel(
        applicationPath,
        distributedFiles,
        applicationPath.resolve("preprocesedApp").toFile,
        networkingOption = Networking.enable)
      f(root)
    }
  }

  private def withTempDirectory[T](f : Path => T) : T = {
    val directory = Files.createTempDirectory("application")
    try {
      f(directory)
    } finally {
      IOUtils.recursiveDeleteDir(directory.toFile)
    }
  }

  private def createServicesXml(applicationPath : Path,
                             containerNode: Node) {

    scala.xml.XML.save(applicationPath.resolve("services.xml").toFile.getAbsolutePath,
      containerNode, xmlDecl = true)
  }
}

File: standalone-container/src/test/java/com/yahoo/container/standalone/StandaloneContainer.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone;

import com.yahoo.collections.Pair;
import com.yahoo.config.model.ConfigModelRepo;
import com.yahoo.io.IOUtils;
import com.yahoo.vespa.model.VespaModel;
import com.yahoo.vespa.model.container.xml.ContainerModelBuilder.Networking;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

/**
 * Creates a local application from vespa-services fragments.
 *
 * @author Tony Vaagenes
 * @author ollivir
 */
public class StandaloneContainer {

    interface ThrowingFunction<T, U> {
        U apply(T input) throws Exception;
    }

    static <T> T withContainerModel(String servicesXml, ThrowingFunction<VespaModel, T> f) throws Exception {
        return withTempDirectory(applicationPath -> {
            writeServicesXml(applicationPath, servicesXml);

            LocalFileDb distributedFiles = new LocalFileDb(applicationPath);
            VespaModel root;
            Pair<VespaModel, com.yahoo.vespa.model.container.Container> rc =
                    StandaloneContainerApplication.createContainerModel(applicationPath, distributedFiles,
                                                                        Networking.enable, new ConfigModelRepo());
            root = rc.getFirst();
            return f.apply(root);
        });
    }

    private static <T> T withTempDirectory(ThrowingFunction<Path, T> f) throws Exception {
        Path directory = Files.createTempDirectory("application");
        try {
            return f.apply(directory);
        } finally {
            IOUtils.recursiveDeleteDir(directory.toFile());
        }
    }

    private static void writeServicesXml(Path applicationPath, String servicesXml) throws IOException {
        Path path = applicationPath.resolve("services.xml");
        List<String> output = List.of("<?xml version=\"1.0\" encoding=\"utf-8\"?>", servicesXml);
        Files.write(path, output, StandardCharsets.UTF_8);
    }
}
